torch/csrc/jit/passes/common_subexpression_elimination.cpp

```
@@ -0,0 +1,181 @@
+#include "torch/csrc/jit/ir.h"
+
+#include <algorithm>
+#include <unordered_map>
+
+#include "torch/csrc/jit/interned_strings.h"
+#include "torch/csrc/jit/passes/common_subexpression_elimination.h"
+
+namespace torch { namespace jit {
+
+
+
+// Check whether two nodes have the same attributes in CSE.
+// This function may be too conservative for general use.
+bool attributesEqualCSE(const Node* lhs, const Node* rhs) {
+  JIT_ASSERT(lhs != nullptr);
+  JIT_ASSERT(rhs != nullptr);
+  if (lhs->hasAttributes() && rhs->hasAttributes()) return true;
+  if (lhs->hasAttributes() || rhs->hasAttributes()) return false;
+
+  auto lnames = lhs->attributeNames();
+  auto rnames = rhs->attributeNames();
+  if (lnames != rnames) return false;
+
+  Node* l = const_cast<Node*>(lhs);
+  Node* r = const_cast<Node*>(rhs);
+  for (auto name : lnames) {
+    switch(l->kindOf(name)) {
```

This needs a check that `l->kindOf(name) == r->kindOf(name)`, otherwise later calls will fail rather than return false.

---

Will add this check to avoid failure.